tools/walletextension/readwriter/read_writer.go

```diff
@@ -79,9 +81,16 @@ func (h *HTTPReadWriter) SupportsSubscriptions() bool {
 	return false
 }
 
+func (h *HTTPReadWriter) IsClosed() bool {
+	return false
+}
+
 func (w *WSReadWriter) ReadRequest() ([]byte, error) {
 	_, msg, err := w.conn.ReadMessage()
 	if err != nil {
+		if websocket.IsCloseError(err) {
```

Comment:
The only way to determine if a websocket is closed is that you receive an "is closed" error when you try and read or write to it. So we check here, so we can update the `ReadWriter`'s internal state and allow others to react accordingly (in this case, cancel existing subscriptions).